go/ct/common/u256_test.go

```diff
@@ -461,3 +462,178 @@ func TestU256_MarshallingRoundTrip(t *testing.T) {
 		}
 	}
 }
+
+func TestU256_RandU256(t *testing.T) {
+	rands := []U256{}
+	rnd := rand.New()
+	for i := 0; i < 10; i++ {
+		rands = append(rands, RandU256(rnd))
+		for j := 0; j < i; j++ {
+			if rands[i] == rands[j] {
+				t.Errorf("random U256 are not random, got %v and %v", rands[i], rands[j])
+			}
+		}
+	}
+}
+
+func TestU256_IsUint64(t *testing.T) {
+	tests := []struct {
+		value U256
+		want  bool
+	}{
+		{U256{}, true},
+		{NewU256(1), true},
+		{NewU256(1, 2), false},
+		{NewU256(1, 2, 3), false},
+		{NewU256(42, 13, 47, 1), false},
+		{NewU256(0xa000000000000000, 0xb000000000000000, 0xc000000000000000, 0xd000000000000000), false},
+		{NewU256(0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff), false},
+	}
+
+	for _, test := range tests {
+		if want, got := test.want, test.value.IsUint64(); want != got {
+			t.Errorf("Unexpected result from IsUint64, want %v, got %v", want, got)
+		}
+	}
+}
+
+func TestU256_Uint64(t *testing.T) {
+	tests := []struct {
+		value U256
+		want  uint64
+	}{
+		{U256{}, 0},
+		{NewU256(1), 1},
+		{NewU256(0xffffffffffffffff), 0xffffffffffffffff},
+	}
+
+	for _, test := range tests {
+		if want, got := test.want, test.value.Uint64(); want != got {
+			t.Errorf("Unexpected result from Uint64, want %v, got %v", want, got)
+		}
+	}
+}
+
+func TestU256_Uint256Conversion(t *testing.T) {
+	tests := []struct {
+		original  *uint256.Int
+		converted U256
+	}{
+		{uint256.NewInt(0), U256{}},
+		{uint256.NewInt(1), NewU256(1)},
+		{uint256.NewInt(256), NewU256(256)},
+		{new(uint256.Int).Lsh(uint256.NewInt(1), 64), NewU256(1, 0)},
+	}
+
+	for _, test := range tests {
+		if want, got := test.original, test.converted.Uint256(); !want.Eq(&got) {
+			t.Errorf("Unexpected result from Uint256, want %v, got %v", want, got)
+		}
+		if want, got := test.converted, NewU256FromUint256(test.original); !want.Eq(got) {
+			t.Errorf("Unexpected result from NewU256FromUint256, want %v, got %v", want, got)
+		}
+	}
+}
+
+func TestU256_SignExtend(t *testing.T) {
+	tests := []struct {
+		value U256
+		want  U256
+	}{
+		{U256{}, U256{}},
+		{NewU256(1), NewU256(1)},
+		{NewU256(0xff), NewU256(0xff)},
+		{NewU256(0xffffffffffffffff), NewU256(0xffffffffffffffff)},
+		{NewU256(0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff), NewU256(0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff)},
+	}
+
+	for _, test := range tests {
+		if want, got := test.want, test.value.SignExtend(NewU256(0xfff)); !want.Eq(got) {
```

Comment:
This test does not cover the case where the argument to `SignExtend` is less than <= 31, in which case the semantic of the function changes.